src/libcore/iter/traits/iterator.rs

```diff
@@ -1859,14 +1859,13 @@ pub trait Iterator {
         Self: Sized, F: FnMut(Self::Item) -> bool
     {
         #[inline]
-        fn check<T>(mut f: impl FnMut(T) -> bool) -> impl FnMut(T) -> LoopState<(), ()> {
-            move |x| {
+        fn check<T>(mut f: impl FnMut(T) -> bool) -> impl FnMut((), T) -> LoopState<(), ()> {
+            move |(), x| {
                 if f(x) { LoopState::Continue(()) }
                 else { LoopState::Break(()) }
             }
         }
-
-        self.try_for_each(check(f)) == LoopState::Continue(())
+        self.try_fold((), check(f)) == LoopState::Continue(())
```

Comment:
Thoughts on equality check vs pattern matching here, can it have an effect or none at all?

Reply:
made a quick diff in https://godbolt.org/z/UW34X_ and there is less code to inline so that is something that I can do
_ZN72_$LT$example..LoopState$LT$C$C$B$GT$$u20$as$u20$core..cmp..PartialEq$GT$2eq17h37dbcaf2df999e09E is a lot to inline

Reply:
I would hope it has no effect, since `LoopState<(),()>` is an `i1` in LLVM...

...and it is in `-O`, but very different in debug: https://rust.godbolt.org/z/LKOpZ7

Looks like the `PartialEq::eq` that gets generated is pretty bad, and it's still bad removing then generics: https://rust.godbolt.org/z/o6Nuaw  Could there be a "this is a field-less enum so just compare the discriminants" path in the derive?  It looks, unfortunately, like `as u8 == 1` is the shortest way to do these checks.  And we're avoiding the derives in other places too, like

https://github.com/rust-lang/rust/blob/702b45e409495a41afcccbe87a251a692b0cefab/src/libcore/cmp.rs#L632-L638

Reply:
Oh interesting. So we could improve it just by implementing PartialEq manually, or even adding a separate method for just discriminant comparison. But then pattern matching works well too.

Reply:
but the pattern match avoids having a function to inline completely as long as a function is used the llvm-ir will contain a call and a function